numpyro/handlers.py

```diff
@@ -391,9 +391,18 @@ def process_message(self, msg):
                     else:
                         msg['value'] = self.base_param_map[msg['name']]
         elif self.substitute_fn is not None:
-            value = self.substitute_fn(msg)
-            if value is not None:
-                msg['value'] = value
+            base_value = self.substitute_fn(msg)
+            if base_value is not None:
+                if msg['type'] == 'sample':
+                    msg['value'], msg['intermediates'] = msg['fn'].transform_with_intermediates(
+                        base_value)
+                else:
+                    constraint = msg['kwargs'].pop('constraint', real)
+                    transform = biject_to(constraint)
+                    if isinstance(transform, ComposeTransform):
+                        msg['value'] = ComposeTransform(transform.parts[1:])(base_value)
```

Comment:
I'm also beginning to think that there was value in our earlier design of also storing `msg['base_value']`, since we are finding more places where we should be working with base values instead of actual values. Nothing related to this PR though. :)

Reply:
Yes, I think so too. Except for computing KL(p||q) in SVI, the current inference algorithms (HMC and autoguide) works in unconstrained space. Hence base value is preferred.

Actually, our approach is similar to [Automatic Reparameterisation of Probabilistic Programs](https://arxiv.org/pdf/1906.03028.pdf) paper. We don't use additional handler for this job and don't rewrite all of our distributions in term of TransformedDistribution. Users can easily trigger non-center reparameterization by defining their priors as a TransformedDistribution (e.g. TransformedDistribution(Normal(0, 1), AffineTransform(loc, scale))`. I just recognize this recently and like our approach better. :)

Reply:
I think that we are having many cool features since the last release. :D

Reply:
Yes, lets start planning a release after jax's next release.